### PR TITLE
Use PersistentMap for cache in brain::Circuit::loadMorphologies()

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

File: brain/circuit.cpp [hunk 1]
```diff
@@ -396,36 +461,65 @@ neuron::Morphologies Circuit::loadMorphologies( const GIDSet& gids,
                                                 const Coordinates coords ) const
 {
     const URIs& uris = getMorphologyURIs( gids );
-    neuron::Morphologies result;
-    result.reserve( uris.size( ));
 
-    if( coords == COORDINATES_GLOBAL )
+    // < GID, hash >
+    boost::unordered_map< size_t, std::string > gidHashes;
```
Comment (hernando) [inline brain/circuit.cpp]: Use a vector instead of an unordered map.

**Comment by tribal-tec** [comment]
retest this please

**Comment by tribal-tec** [comment]
ping

File: brain/circuit.cpp [hunk 2]
```diff
@@ -396,36 +455,63 @@ neuron::Morphologies Circuit::loadMorphologies( const GIDSet& gids,
...
+
+        if( coords == COORDINATES_GLOBAL )
         {
-            const URI& uri = uris[i];
-            const brion::Morphology raw( uri.getPath( ));
-            result.push_back( neuron::MorphologyPtr(
-                                 new neuron::Morphology( raw, transforms[i] )));
+            // store circuit + GID for transformed morphology
+            hash += _impl->getCircuitSource().getPath() +
+                       boost::lexical_cast< std::string >(i+1);
```
Comment (hernando) [inline brain/circuit.cpp]: You have to iterate over the GIDs at the same time you loop because i is just an index.

Comment (tribal-tec) [inline brain/circuit.cpp]: Good catch!

**Comment by tribal-tec** [comment]
Updated

File: brain/circuit.cpp [hunk 3]
```diff
@@ -396,36 +459,64 @@ neuron::Morphologies Circuit::loadMorphologies( const GIDSet& gids,
                                                 const Coordinates coords ) const
 {
     const URIs& uris = getMorphologyURIs( gids );
```
Comment (hernando) [inline brain/circuit.cpp]: Can you double check what does this function do when a gid doesn't exist in both mvd2 and mvd3?

File: brain/circuit.cpp [hunk 4]
```diff
@@ -396,36 +459,64 @@ neuron::Morphologies Circuit::loadMorphologies( const GIDSet& gids,
...
+    Strings gidHashes;
+    gidHashes.reserve( uris.size( ));
+    std::set< std::string > hashes;
+    GIDSet::const_iterator gid = gids.begin();
+    for( size_t i = 0; i < uris.size(); ++i, ++gid )
     {
-        const Matrix4fs& transforms = getTransforms( gids );
-        for( size_t i = 0; i < uris.size(); ++i )
+        const URI& uri = uris[i];
+        std::string hash( uri.getPath( ));
```
Comment (hernando) [inline brain/circuit.cpp]: Merge with line above.

File: brain/neuron/morphologyImpl.cpp [hunk 5]
```diff
@@ -51,6 +56,98 @@ Morphology::Impl::Impl( const brion::Morphology& morphology )
...
+bool Morphology::Impl::_fromBinary( const void* data, const size_t size )
+{
+    const uint8_t* ptr = reinterpret_cast< const uint8_t* >( data );
+
+    size_t arraySize = *reinterpret_cast< const size_t* >( ptr );
+    ptr += sizeof(size_t);
+    const brion::Vector4f* pointPtr =
+            reinterpret_cast< const brion::Vector4f* >( ptr );
+    points.reset( new brion::Vector4fs( pointPtr, pointPtr + arraySize ));
+    ptr += sizeof(brion::Vector4f) * arraySize;
```
Comment (hernando) [inline brain/neuron/morphologyImpl.cpp]: Write a template function to not repeat the same code three times?

File: brain/neuron/morphologyImpl.cpp [hunk 6]
```diff
@@ -51,6 +56,98 @@ Morphology::Impl::Impl( const brion::Morphology& morphology )
...
+
+    uint8_t* ptr = new uint8_t[data.size];
+    data.ptr.reset( ptr );
+
+    // points
+    size_t arraySize = points->size();
+    *reinterpret_cast< size_t* >( ptr ) = arraySize;
+    ptr += sizeof(size_t);
+    memcpy( ptr, points->data(), sizeof(brion::Vector4f) * points->size());
+    ptr += sizeof(brion::Vector4f) * points->size();
```
Comment (hernando) [inline brain/neuron/morphologyImpl.cpp]: Same here

File: brion/plugin/morphologyHDF5.cpp [hunk 7]
```diff
@@ -760,6 +760,7 @@ bool MorphologyHDF5:: _readV2Metadata() const
     {
         detail::SilenceHDF5 silence;
         _file.openGroup( _g_root );
+        _version = MORPHOLOGY_VERSION_H5_2;
```
Comment (hernando) [inline brion/plugin/morphologyHDF5.cpp]: This means it's possible to have v2 morphology files without the version attribute?

File: doc/Changelog.md [hunk 8]
```diff
@@ -1,6 +1,11 @@
 Changelog {#Changelog}
 =========
 
+# git master
+
+* [79](https://github.com/BlueBrain/Brion/pull/79):
+  Use PersistentMap for cache in brain::Circuit::loadMorphologies()
```
Comment (hernando) [inline doc/Changelog.md]: And synapses cache access thread safety.

File: brain/circuit.cpp (hunk 3 above)
Comment (tribal-tec) [inline brain/circuit.cpp]: See https://github.com/BlueBrain/Brion/pull/81

File: brain/neuron/morphologyImpl.cpp (hunk 5 above)
Comment (tribal-tec) [inline brain/neuron/morphologyImpl.cpp]: Or use Zerobuf :)

File: brion/plugin/morphologyHDF5.cpp (hunk 7 above)
Comment (tribal-tec) [inline brion/plugin/morphologyHDF5.cpp]: According to the code, yes. It's still the same which was initially implemented in the SDK years ago.

**Comment by tribal-tec** [comment]
Updated

File: brain/circuit.cpp [hunk 9]
```diff
@@ -146,10 +159,60 @@ class Circuit::Impl
...
+    }
+
+    Loaded loadFromCache( const std::set< std::string >& hashes LB_UNUSED )
+        const
+    {
+        Loaded loaded;
+#ifdef BRION_USE_CXX11
+        if( _cache )
+        {
+            LBINFO << "Using cache for morphology loading" << std::endl;
```
Comment (eile) [inline brain/circuit.cpp]: LBDEBUG?

File: brain/circuit.cpp [hunk 10]
```diff
@@ -396,36 +506,63 @@ neuron::Morphologies Circuit::loadMorphologies( const GIDSet& gids,
...
+             coords == COORDINATES_GLOBAL ? getTransforms( gids ) : Matrix4fs();
     for( size_t i = 0; i < uris.size(); ++i )
     {
         const URI& uri = uris[i];
 
-        neuron::MorphologyPtr& morphology = loaded[uri.getPath()];
-        if( !morphology )
+        const std::string& hash = gidHashes[i];
+        Loaded::const_iterator it = loaded.find( hash );
+        if( it != loaded.end( ))
```
Comment (eile) [inline brain/circuit.cpp]: swap condition for ==?

File: brain/neuron/morphologyImpl.cpp [hunk 11]
```diff
@@ -51,6 +82,50 @@ Morphology::Impl::Impl( const brion::Morphology& morphology )
...
+    _extractChildrenLists();
+
+    const uint32_ts ids =
+        getSectionIDs( SectionTypes( 1, SECTION_SOMA ), false );
+
+    if( ids.size() != 1 )
+        LBTHROW( std::runtime_error(
+                    "Bad input morphology. None or more than one soma found" ));
+    somaSection = ids[0];
+    return true;
```
Comment (eile) [inline brain/neuron/morphologyImpl.cpp]: yuck

Comment (tribal-tec) [inline brain/neuron/morphologyImpl.cpp]: ?

Comment (eile) [inline brain/neuron/morphologyImpl.cpp]: nice code :-/

**Comment by tribal-tec** [comment]
@rdumusc Please review at least the caching.md if it's understandable for you :)

File: doc/caching.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+* MEMCACHED_SERVERS: a comma-separated list of servers with optional :port to
+                     use memcached as a cache
+* LEVELDB_CACHE: a path to the leveldb storage to use leveldb as a cache
+
+## Cached data support
+
+* Morphologies from brain::Circuit::loadMorphologies()
+  * each morphology is hashed by its absolute filepath if COORDINATES_LOCAL
+  * each morphology is hashed by its absolute filepath plus circuit filepath and
+    GID if COORDINATES_GLOBAL
```
Comment (eile) [inline doc/caching.md]: Shouldn't this use the canonical path too?

**Comment by rdumusc** [comment]
+1, I find doc/caching.md very clear to read. Based on this, I understand that the cluster module will set the appropriate variables.

**Pull request merged**